docs/specification/2025-03-26/basic/authorization.mdx

```diff
@@ -8,303 +8,497 @@ title: Authorization
+MCP servers **MUST NOT** complete public client flows on behalf of the MCP client (e.g., take 
+authorization codes and exchange them for tokens with a public client ID, and then pass the tokens 
+back to the client).
+
+#### 2.4.2 Flow Description
+
+The third-party authorization flow comprises these steps:
+
+1. MCP client requests data from the MCP server.  
+1. MCP server responds with `HTTP 401 Unauthorized` and `WWW-Authenticate` header containing the
+   URL to the PRM document. 
+1. MCP client retrieves and validates the PRM document.
+1. MCP client retrieves the AS endpoint from the PRM document.  
+1. MCP client redirects user to the AS authorization endpoint. MCP client can include scopes that
+   are defined in the PRM document or other scopes, as supported by the MCP server.
+1. User authorizes with authorization server.
+1. Authorization server redirects back to MCP client with authorization code.  
+1. MCP client exchanges the code for the requested tokens.  
+1. MCP client stores the token.  
+1. MCP client can now request data from the MCP server with the token attached in the `Authorization`
+   header.
 
 ```mermaid
 sequenceDiagram
-    participant C as Client
-    participant S as Server
-
-    C->>S: GET /.well-known/oauth-authorization-server
-    alt Discovery Success
-        S->>C: 200 OK + Metadata Document
-        Note over C: Use endpoints from metadata
-    else Discovery Failed
-        S->>C: 404 Not Found
-        Note over C: Fall back to default endpoints
-    end
-    Note over C: Continue with authorization flow
+    participant UA as User-Agent (Browser)
+    participant MC as MCP Client
+    participant MS as MCP Server
+    participant AS as Authorization Server
+
+    MC->>MS: Request data
+    MS-->>MC: 401 Unauthorized with WWW-Authenticate
+    Note over MC: MCP Client retrieves and validates<br/>PRM document
+    Note over MC: MCP Client retrieves authorization<br/>endpoint from AS metadata document<br/>(Section 2.3)
+    MC-->>UA: Initiate authorization request at the authorization endpoint
+    Note over UA: User authenticates and authorizes<br/>client for scope and resource
+    UA->>MC: Return authorization code to client
+    MC->>AS: Exchange code for token
+    AS-->>MC: Access token
+    MC->>MC: Cache token
 ```
 
-#### 2.3.1 Server Metadata Discovery Headers
+#### 2.4.4 Security Considerations
+
+When implementing authorization with an existing AS, MCP servers **MUST**:
+
+1. Request the minimum number of required scopes.  
+2. Securely store any tokens obtained from the client and/or the authorization server.  
+3. Implement appropriate session timeout handling.  
+4. Consider security implications of token chaining.  
+5. Implement proper error handling for authorization failures.
+
+MCP clients **MUST**:
+
+1. Securely store Access and Refresh Tokens received from the AS.  
+2. Properly handle requests to re-authenticate users with expired or invalid tokens (using the
```

Comment:
The refresh token grant is not re-authenticating the user, this should be reworded.

Reply:
Updated.